format-currency: Add isSmallestUnit option to support integer amounts

sirbrillig opened the pull request:
#### Proposed Changes

The `@automattic/format-currency` package handles formatting floating point numbers like `2345.23` for a locale like `$2,345.23`. However, since rounding errors in floating point math are common, most payment systems use integers like `234523` for a currency amount, keeping the number in the currency's smallest unit (eg: cents for USD, yen for JPY).

This PR adds a new option to the functions exported by the `format-currency` package: `isSmallestUnit?: boolean`. If true, the functions will assume that the amount is an integer instead of a floating point number and convert it for display appropriately.

This may be needed by https://github.com/Automattic/wp-calypso/pull/66066 but also could come in handy in other places.

#### Testing Instructions

Automated tests are included.

sirbrillig committed:
Add smallestUnit to formatCurrency

sirbrillig committed:
Export types from format-currency package

sirbrillig committed:
Rename option to isSmallestUnit

manzoorwanijk commented on packages/format-currency/src/index.ts:
```diff
@@ -6,17 +6,20 @@ export { getCurrencyDefaults };
 
 export { CURRENCIES } from './currencies';
 
+export * from './types';
```
Thank you for this. I wanted to do this in some PR.

tyxla commented on packages/format-currency/src/index.ts:
```diff
@@ -6,17 +6,20 @@ export { getCurrencyDefaults };
  * Formats money with a given currency code
  *
- * @param   {number}     number              number to format
+ * @param   {number}     number              number to format; assumed to be a float unless isSmallestUnit is set.
  * @param   {string}     code                currency code e.g. 'USD'
  * @param   {object}     options             options object
  * @param   {string}     options.decimal     decimal symbol e.g. ','
  * @param   {string}     options.grouping    thousands separator
  * @param   {number}     options.precision   decimal digits
  * @param   {string}     options.symbol      currency symbol e.g. 'A$'
  * @param   {boolean}    options.stripZeros  whether to remove trailing zero cents
+ * @param   {boolean}    options.isSmallestUnit  true if the number is an integer in the currency's smallest unit.
```
Nit: Prior to this PR, the docblock sections were aligned. Should we maintain that?

```suggestion
 * @param   {number}     number                  number to format; assumed to be a float unless isSmallestUnit is set.
 * @param   {string}     code                    currency code e.g. 'USD'
 * @param   {object}     options                 options object
 * @param   {string}     options.decimal         decimal symbol e.g. ','
 * @param   {string}     options.grouping        thousands separator
 * @param   {number}     options.precision       decimal digits
 * @param   {string}     options.symbol          currency symbol e.g. 'A$'
 * @param   {boolean}    options.stripZeros      whether to remove trailing zero cents
 * @param   {boolean}    options.isSmallestUnit  true if the number is an integer in the currency's smallest unit.
```

manzoorwanijk replied:
May be this?
```suggestion
 * @param   {boolean}    options.isSmallestUnit  whether the number is an integer in the currency's smallest unit.
```

Same for the other one.

manzoorwanijk reviewed (approved):
Looks good. I am glad the package is shaping up nicely. May be it's the time to add some documentation to it in some other PR?

tyxla commented on packages/format-currency/src/index.ts:
```diff
@@ -28,8 +31,17 @@ export default function formatCurrency(
 	const sign = number < 0 ? '-' : '';
+	if ( isSmallestUnit ) {
+		if ( ! Number.isInteger( number ) ) {
+			return null;
```
We could use a unit test to cover that scenario.

tyxla commented on packages/format-currency/src/index.ts:
```diff
@@ -60,8 +73,17 @@ export function getCurrencyObject(
 	const sign = number < 0 ? '-' : '';
+	if ( isSmallestUnit ) {
+		if ( ! Number.isInteger( number ) ) {
+			return null;
```
We could use a unit test to cover that scenario.

tyxla commented on packages/format-currency/src/index.ts:
```diff
@@ -28,8 +31,17 @@ export default function formatCurrency(
 		return null;
 	}
-	const { decimal, grouping, precision, symbol } = { ...currencyDefaults, ...options };
+	const { decimal, grouping, precision, symbol, isSmallestUnit } = {
```
I wonder if we could figure out a better name than `isSmallestUnit`... Is there a generic term for the coin of a currency that we could use here? `coin` is one, but I guess it also adds some ambiguity?

Also, are there currencies with more than one smaller unit than the primary one? I'd trust y'all payments folks would be aware if there's one but just wanted to make sure it's been considered.

sirbrillig replied:
Stripe calls it [smallest currency unit](https://stripe.com/docs/currencies#zero-decimal), which is where I got the term. Probably because of that we call it the "smallest unit" on our backend, although in the shopping cart we just call it "integer" (eg: `original_price_integer`). I did consider calling this `isInteger` but I felt that was probably even more ambiguous than `isSmallestUnit`. 🤷

tyxla replied:
Thanks for confirming 👍 Naming is hard 😅

tyxla reviewed (approved):
This looks great! 

I've left some comments and questions, but everything is optional and up to you to address before merging if you prefer.

Otherwise, 🚀

danielmx-dev commented on packages/format-currency/src/index.ts:
```diff
@@ -28,8 +31,17 @@ export default function formatCurrency(
+		if ( ! Number.isInteger( number ) ) {
+			return null;
+		}
+		number = convertPriceForSmallestUnit( number, precision );
```
What would happen if you override the `precision` value in the `options` object? Would it return an inaccurate result? If that's the case maybe the `currencyDefaults.precision` value should be used here to prevent that scenario.

sirbrillig replied:
🤔 It would return an accurate result considering the new `precision`, but I see your point. `precision` may have two implied meanings here: the number of decimals used by the currency and the number of decimal places we want to see rendered. 

I _think_ (and I may be wrong) that the intent of the `precision` argument is primarily to override the _currency's_ precision rather than to control how many decimals are displayed because we also have the `stripZeros` option. I suspect that with that option there's no need to actually change `precision` itself. That said, the documentation should clarify that and I'll see if I can do that in this PR.

The only place I think that we're currently overriding `precision` is [here](https://github.com/Automattic/wp-calypso/blob/5508aab1fba60b93fb785b042e422042647d0673/client/my-sites/checkout/composite-checkout/hooks/product-variants.tsx#L39), and I think that `stripZeros` might work better there (but I'll investigate further to be sure).

sirbrillig committed:
Reformat docblock alignment

sirbrillig committed:
Document each option

sirbrillig committed:
Provide details in each function's documentation

Removes the option JSDoc documentation in favor of the TS documentation
and the function's prose.

sirbrillig committed:
Add docs for all exports to README

sirbrillig committed:
Clarify type of CurrencyObject.sign

sirbrillig committed:
Add tests for isSmallestUnit and floats

sirbrillig commented:
> May be it's the time to add some documentation to it in some other PR?

I decided to just improve the docs here because it was easier to explain the new option and the perils of setting `precision` that @dylanrevisited brought up that way.
